scripts/validate_rst_title_capitalization.py

```diff
@@ -0,0 +1,306 @@
+        RST file to scan through for headings
+
+    Returns
+    -------
+    title_list : List[str]
+        A list of heading strings found in the document tree
+
+    line_number_list : List[int]
+        The corresponding line numbers of the headings in title_list
+
+    """
+
+    # title_list is the list of headings that is encountered in the doctree
+    title_list: List[str] = []
+
+    # List of line numbers that corresponding headings in title_list can be found at
+    line_number_list: List[int] = []
+
+    # Open and read the .rst file and store the string of data into input
+    f = open(rst_file, "r")
+    input = f.read().split("\n")
+    f.close()
+
+    # Regular expressions that denote a title beforehand
+    regex = {
+        "*": r"^(?:\*{1})*$",
+        "=": r"^(?:={1})*$",
+        "-": r"^(?:-{1})*$",
+        "^": r"^(?:\^{1})*$",
+        "~": r"^(?:~{1})*$",
+        "#": r"^(?:#{1})*$",
+        '"': r'^(?:"{1})*$',
+    }
+
+    # '*`_' markers are removed from original string text.
+    table = str.maketrans("", "", "*`_")
+
+    # Loop through input lines, appending if they are considered headings
+    for lineno in range(1, len(input)):
+        if len(input[lineno]) != 0 and len(input[lineno - 1]) != 0:
+            for key in regex:
+                match = re.search(regex[key], input[lineno])
+                if match is not None:
+                    if lineno >= 2:
+                        if input[lineno] == input[lineno - 2]:
+                            if len(input[lineno]) == len(input[lineno - 1]):
+                                title_list.append(input[lineno - 1].translate(table))
+                                line_number_list.append(lineno)
+                            break
+                    if len(input[lineno]) >= len(input[lineno - 1]):
+                        title_list.append(input[lineno - 1].translate(table))
+                        line_number_list.append(lineno)
+
+    return title_list, line_number_list
+
+
+def fill_bad_title_dict(rst_file: str) -> None:
+    """
+    Method that fills up the bad_title_dict with incorrectly capitalized headings
+
+    Parameters
+    ----------
+    rst_file : str
+        Directory address of a .rst file as a string
+
+    """
+
+    # Ensure this file doesn't already have a bad_title_dict slot
+    if rst_file in bad_title_dict:
+        return
+
+    # Make a list of headings along with their line numbers
+    title_list, line_number_list = findTitles(rst_file)
+
+    # Append the bad_title_dict if the capitalization convention not followed
+    for i in range(len(title_list)):
+        if not follow_capitalization_convention(title_list[i]):
+            if rst_file not in bad_title_dict:
+                bad_title_dict[rst_file] = [(title_list[i], line_number_list[i])]
+            else:
+                bad_title_dict[rst_file].append((title_list[i], line_number_list[i]))
+
+
+def find_rst_files(source_paths: List[str]) -> List[str]:
```

Comment:
I think that the return type here is:

```python
Generator[str, None, None]
```

Reply:
@MomIsBestFriend Thank you for your input!  It's really helping me a lot.

Reply:
No problems, and if I gotta be honest I learnt a lot from @datapythonista, I would suggest you to read his code suggestions he gave me in https://github.com/pandas-dev/pandas/pull/30467 and in https://github.com/pandas-dev/pandas/pull/30755

I think that I stepped in pretty much every possible pitfall there is.